samples/android/app/src/main/java/org/hyperledger/indy/sdk/ui/ledger/LedgerActivity.kt

```diff
@@ -0,0 +1,314 @@
+import org.hyperledger.indy.sdk.did.DidJSONParameters
+import org.hyperledger.indy.sdk.helpers.MessageHelper
+import org.hyperledger.indy.sdk.ledger.Ledger
+import org.hyperledger.indy.sdk.pool.Pool
+import org.hyperledger.indy.sdk.utils.PoolUtils
+import org.hyperledger.indy.sdk.wallet.Wallet
+import org.json.JSONObject
+import org.junit.Assert
+import java.lang.Exception
+
+class LedgerActivity : AppCompatActivity() {
+
+    private val TAG = LedgerActivity::class.java.name
+
+
+    // my vars
+    private val trusteeSeed = "000000000000000000000000Trustee1"
+    private lateinit var myWallet: Wallet
+    private lateinit var trusteeWallet: Wallet
+    private lateinit var trusteeDid: String
+    private lateinit var myDid: String
+    private lateinit var myVerkey: String
+    private lateinit var pool: Pool
+    private lateinit var nymRequest: String
+    private lateinit var myWalletConfig: String
+    private lateinit var myWalletCredentials: String
+    private lateinit var trusteeWalletConfig: String
+    private lateinit var trusteeWalletCredentials: String
+    private lateinit var poolName: String
+
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_ledger)
+
+        // start Ledger demo
+        startDemo()
+    }
+
+    private fun updateUI(text: String) {
+        tvLedgerLogs.text = "${tvLedgerLogs.text}$text"
+    }
+
+    private fun updateHeader(text: String) {
+        pbLedger.visibility = View.VISIBLE
+        tvLedgerStart.text = text
+    }
+
+    private fun updateFooter(text: String) {
+        pbLedger.visibility = View.GONE
+        tvLedgerEnd.text = text
+    }
+
+    /**
+     * startDemo function start all functions for Ledger demo chronological in coroutine default thread
+     */
+    private fun startDemo() {
+        MainScope().launch {
+            Log.d(TAG, "startDemo: Ledger sample -> STARTED!")
+
+            updateHeader(getString(R.string.ledger_sample_start))
+
+
+            updateUI(getString(R.string.ledger_create_ledger))
+            try {
+                withContext(CoroutineScope(Dispatchers.Default).coroutineContext) {
+                    createLedger()
+                }
+            } catch (e: Exception) {
+                MessageHelper.errorToast(this@LedgerActivity, getString(R.string.error))
+                pbLedger.visibility = View.GONE
+                return@launch
+            }
+            updateUI(getString(R.string.ledger_create_ledger_end))
+
+            updateUI(getString(R.string.ledger_create_open_my_wallet))
+            try {
+                withContext(CoroutineScope(Dispatchers.Default).coroutineContext) {
+                    createOpenMyWallet()
+                }
+            } catch (e: Exception) {
+                MessageHelper.errorToast(this@LedgerActivity, getString(R.string.error))
+                pbLedger.visibility = View.GONE
+                return@launch
```

Comment:
you are returning after each error. why wouldn't you just propagate that error one level higher and handle it there.

also, I would write errors just as 1 line in the `updateUi` method. maybe only color it red.